lib/web/apiserver_login_test.go

```diff
@@ -337,6 +346,158 @@ func TestAuthenticate_rateLimiting(t *testing.T) {
 	}
 }
 
+func TestAuthenticate_deviceWebToken(t *testing.T) {
+	pack := newWebPack(t, 1 /* numProxies */)
+	authServer := pack.server.AuthServer.AuthServer
+	proxy := pack.proxies[0]
+	clock := pack.clock
+
+	// Mimic a successfully create DeviceWebToken, regarldess of any parameters.
+	wantToken := &types.DeviceWebToken{
+		Id:    "this is an opaque token ID",
+		Token: "this is an opaque token Token",
+	}
+	authServer.SetCreateDeviceWebTokenFunc(func(context.Context, *devicepb.DeviceWebToken) (*devicepb.DeviceWebToken, error) {
+		return &devicepb.DeviceWebToken{
+			Id:    wantToken.Id,
+			Token: wantToken.Token,
+		}, nil
+	})
+
+	ctx := context.Background()
+
+	t.Run("login using OTP", func(t *testing.T) {
+		// Create a user with password + OTP.
+		const user = "llama1"
+		const pass = "mysupersecretpassword!!1!"
+		otpSecret := newOTPSharedSecret()
+		proxy.createUser(ctx, t, user, user, pass, otpSecret, nil /* roles */)
+
+		sessionResp := loginWebOTP(t, ctx, loginWebOTPParams{
+			webClient: proxy.newClient(t),
+			clock:     clock,
+			user:      user,
+			password:  pass,
+			otpSecret: otpSecret,
+		})
+		assert.Equal(t, wantToken, sessionResp.DeviceWebToken, "WebSession DeviceWebToken mismatch")
+	})
+
+	t.Run("login using WebAuthn", func(t *testing.T) {
+		rpID := pack.server.TLS.ClusterName()
+		mfaResp := configureClusterForMFA(t, pack, &types.AuthPreferenceSpecV2{
+			Type:         constants.Local,
+			SecondFactor: constants.SecondFactorOptional,
+			Webauthn: &types.Webauthn{
+				RPID: rpID,
+			},
+		})
+
+		sessionResp := loginWebMFA(ctx, t, loginWebMFAParams{
+			webClient:     proxy.newClient(t),
+			rpID:          rpID,
+			user:          mfaResp.User,
+			password:      mfaResp.Password,
+			authenticator: mfaResp.WebDev.Key,
+		})
+		assert.Equal(t, wantToken, sessionResp.DeviceWebToken, "WebSession DeviceWebToken mismatch")
+	})
+}
+
+// newOTPSharedSecret returns an OTP shared secret, encoded as a base32 string.
+func newOTPSharedSecret() string {
+	return base32.StdEncoding.EncodeToString([]byte("supersecretsecret!!1!"))
+}
+
+type loginWebOTPParams struct {
+	webClient      *TestWebClient
+	clock          clockwork.Clock
+	user, password string
+	otpSecret      string // base32-encoded shared OTP secret
+}
+
+// loginWebOTP logins the user using the /webapi/sessions/new endpoint.
+//
+// This is a lower-level utility for tests that want access to the returned
+// CreateSessionResponse.
+func loginWebOTP(t *testing.T, ctx context.Context, params loginWebOTPParams) *CreateSessionResponse {
```

Comment:
This and loginWebMFA were put together by stealing bits from various other tests. We don't really have good, standalone helpers that operate at this level of abstraction (some are tied to webPack, others to WebSuite, various tests roll out their own, etc).

I might follow up tomorrow with some testing refactors so we use these more widely (and substitute the many variants popping around).

Reply:
Yeah there are fair amount of authenticated user setup in `oss` spread across test cases. 🎖️🫡 if you could come up with a reusable standard-ish test helper.

Reply:
It is, unsurprisingly, kinda hard to unentangle. I have a draft PR branched from this that I started yesterday. I think I'll have to change the function signatures a bit and maybe provide a few "overloaded" variants, but it's cutting a bit of the repetition.

I'll CC you when I mail it.

Reply:
Oh, lib/web only. I'm not brave enough to tackle _all_ of OSS.